File: patterns/_tests/BehavioralPatterns/Observer/csharp/integration.test.cs
// ============================================================
//  Integration Tests — Observer: Weather Station (C#)
//
//  What we test here:
//    · Full workflow with multiple observers
//    · Correct behavior across subscription lifecycle
//    · Data consistency across all observers
//    · Real-world scenario with mixed operations
//
//  Run: dotnet test
// ============================================================

using System;
using System.Collections.Generic;
using Xunit;


class WeatherUpdate
{
    public double Temperature { get; set; }
    public double Humidity { get; set; }
    public double Pressure { get; set; }
    public double WindSpeed { get; set; }
}

class TrackingObserver : Observer
{
    public List<WeatherUpdate> Updates { get; } = new();

    public void Update(WeatherStation station)
    {
        var reading = station.GetCurrentReading();
        Updates.Add(new WeatherUpdate
        {
            Temperature = reading.Temperature,
            Humidity = reading.Humidity,
            Pressure = reading.Pressure,
            WindSpeed = reading.WindSpeed,
        });
    }
}


public class IntegrationTest
{
    private WeatherStation _station;
    private TemperatureChart _tempChart;
    private HumidityChart _humidityChart;
    private PressureWindChart _pressureChart;
    private TrackingObserver _tracker;

    public IntegrationTest()
    {
        _station = new WeatherStation();
        _tempChart = new TemperatureChart("Temp");
        _humidityChart = new HumidityChart("Humidity");
        _pressureChart = new PressureWindChart("Pressure");
        _tracker = new TrackingObserver();
    }

    [Fact]
    public void TestMultipleObserversTrackMultipleUpdates()
    {
        _station.RegisterObserver(_tempChart);
        _station.RegisterObserver(_humidityChart);
        _station.RegisterObserver(_pressureChart);
        _station.RegisterObserver(_tracker);

        _station.SetWeatherData(18, 45, 1015, 10);
        _station.SetWeatherData(22, 60, 1010, 8);
        _station.SetWeatherData(25, 70, 1015, 12);

        Assert.Equal(3, _tracker.Updates.Count);
        Assert.Equal(18, _tracker.Updates[0].Temperature);
        Assert.Equal(22, _tracker.Updates[1].Temperature);
        Assert.Equal(25, _tracker.Updates[2].Temperature);
    }

    [Fact]
    public void TestUnsubscribeMidStreamStopsUpdates()
    {
        _station.RegisterObserver(_tracker);

        _station.SetWeatherData(18, 45, 1015, 10);
        Assert.Equal(1, _tracker.Updates.Count);

        _station.RemoveObserver(_tracker);

        _station.SetWeatherData(22, 60, 1010, 8);
        Assert.Equal(1, _tracker.Updates.Count); // No new update

        _station.RegisterObserver(_tracker);

        _station.SetWeatherData(25, 70, 1015, 12);
        Assert.Equal(2, _tracker.Updates.Count); // Resumes after re-subscribe
    }

    [Fact]
    public void TestAllObserversSeesSameCurrentReading()
    {
        var tracker1 = new TrackingObserver();
        var tracker2 = new TrackingObserver();
        var tracker3 = new TrackingObserver();

        _station.RegisterObserver(tracker1);
        _station.RegisterObserver(tracker2);
        _station.RegisterObserver(tracker3);

        _station.SetWeatherData(25, 70, 1015, 15);

        var u1 = tracker1.Updates[0];
        var u2 = tracker2.Updates[0];
        var u3 = tracker3.Updates[0];

        Assert.Equal(u1.Temperature, u2.Temperature);
        Assert.Equal(u2.Temperature, u3.Temperature);
        Assert.Equal(u1.Humidity, u2.Humidity);
        Assert.Equal(u1.Pressure, u2.Pressure);
    }

    [Fact]
    public void TestReadingHistoryRemainConsistentAcrossObserverLifespan()
    {
        var tracker1 = new TrackingObserver();

        _station.RegisterObserver(tracker1);
        _station.SetWeatherData(18, 45, 1015, 10);
        _station.SetWeatherData(22, 60, 1010, 8);

        // Get history at first observation point
        var history1 = _station.GetReadings();

        _station.SetWeatherData(25, 70, 1015, 12);

        // Get history after another update
        var history2 = _station.GetReadings();

        Assert.True(history1.Count < history2.Count);
        Assert.Equal(history1[0].Temperature, history2[0].Temperature);
        Assert.Equal(history1[1].Temperature, history2[1].Temperature);
    }

    [Fact]
    public void TestWorksWithMixedObserverSubscribeUnsubscribeSequence()
    {
        var tracker1 = new TrackingObserver();
        var tracker2 = new TrackingObserver();
        var tracker3 = new TrackingObserver();

        // Day 1: Only tracker1
        _station.RegisterObserver(tracker1);
        _station.SetWeatherData(18, 45, 1015, 10);
        Assert.Equal(1, tracker1.Updates.Count);
        Assert.Equal(0, tracker2.Updates.Count);
        Assert.Equal(0, tracker3.Updates.Count);

        // Day 2: Add tracker2
        _station.RegisterObserver(tracker2);
        _station.SetWeatherData(22, 60, 1010, 8);
        Assert.Equal(2, tracker1.Updates.Count);
        Assert.Equal(1, tracker2.Updates.Count);
        Assert.Equal(0, tracker3.Updates.Count);

        // Day 3: Add tracker3, remove tracker1
        _station.RegisterObserver(tracker3);
        _station.RemoveObserver(tracker1);
        _station.SetWeatherData(25, 70, 1015, 12);
        Assert.Equal(2, tracker1.Updates.Count); // Stopped
        Assert.Equal(2, tracker2.Updates.Count);
        Assert.Equal(1, tracker3.Updates.Count);

        // Day 4: Re-add tracker1
        _station.RegisterObserver(tracker1);
        _station.SetWeatherData(28, 75, 1020, 14);
        Assert.Equal(3, tracker1.Updates.Count); // Resumed
        Assert.Equal(3, tracker2.Updates.Count);
        Assert.Equal(2, tracker3.Updates.Count);
    }

    [Fact]
    public void TestRespectsLast10ReadingsHistoryLimit()
    {
        var tracker = new TrackingObserver();
        _station.RegisterObserver(tracker);

        // Generate 15 readings
        for (int i = 0; i < 15; i++)
        {
            _station.SetWeatherData(10 + i, 50 + i, 1010 + i, 10);
        }

        // Should have exactly 10 readings in history
        var readings = _station.GetReadings();
        Assert.Equal(10, readings.Count);

        // Should be the last 10 updates
        Assert.Equal(15, readings[0].Temperature);
        Assert.Equal(24, readings[9].Temperature);
    }

    [Fact]
    public void TestSupportsArbitraryObserverImplementations()
    {
        var avgObserver = new TempAverageObserver();
        _station.RegisterObserver(avgObserver);

        _station.SetWeatherData(20, 60, 1010, 10);
        _station.SetWeatherData(24, 65, 1015, 12);
        _station.SetWeatherData(26, 70, 1020, 15);

        var avg = avgObserver.GetAverage();
        Assert.Equal((20 + 24 + 26) / 3.0, avg, 2);
    }

    [Fact]
    public void TestAllThreeConcreteObserversCoexistWithoutInterference()
    {
        var tracker = new TrackingObserver();

        _station.RegisterObserver(_tempChart);
        _station.RegisterObserver(_humidityChart);
        _station.RegisterObserver(_pressureChart);
        _station.RegisterObserver(tracker);

        // Perform updates
        _station.SetWeatherData(18, 45, 1015, 10);
        _station.SetWeatherData(25, 70, 1020, 15);

        // All should have been notified
        Assert.Equal(2, tracker.Updates.Count);

        // Reading history accessible to all
        var readings = _station.GetReadings();
        Assert.True(readings.Count > 2);
    }
}


public class TempAverageObserver : Observer
{
    private double _tempSum = 0;
    private int _count = 0;

    public void Update(WeatherStation station)
    {
        var current = station.GetCurrentReading();
        _tempSum += current.Temperature;
        _count++;
    }

    public double GetAverage()
    {
        return _count > 0 ? _tempSum / _count : 0;
    }
}

File: patterns/_tests/BehavioralPatterns/Observer/java/integration.test.java
// ============================================================
//  Integration Tests — Observer: Weather Station (Java)
//
//  What we test here:
//    · Full workflow with multiple observers
//    · Correct behavior across subscription lifecycle
//    · Data consistency across all observers
//    · Real-world scenario with mixed operations
//
//  Run: javac -cp .:junit-4.13.2.jar:hamcrest-core-1.3.jar ObserverIntegrationTest.java && java -cp .:junit-4.13.2.jar:hamcrest-core-1.3.jar org.junit.runner.JUnitCore ObserverIntegrationTest
// ============================================================

import org.junit.*;
import static org.junit.Assert.*;
import java.util.*;


public class IntegrationTest {
    private WeatherStation station;
    private TemperatureChart tempChart;
    private HumidityChart humidityChart;
    private PressureWindChart pressureChart;
    private TrackingObserver tracker;

    @Before
    public void setUp() {
        station = new WeatherStation();
        tempChart = new TemperatureChart("Temp");
        humidityChart = new HumidityChart("Humidity");
        pressureChart = new PressureWindChart("Pressure");
        tracker = new TrackingObserver();
    }

    @Test
    public void testMultipleObserversTrackMultipleUpdates() {
        station.registerObserver(tempChart);
        station.registerObserver(humidityChart);
        station.registerObserver(pressureChart);
        station.registerObserver(tracker);

        station.setWeatherData(18, 45, 1015, 10);
        station.setWeatherData(22, 60, 1010, 8);
        station.setWeatherData(25, 70, 1015, 12);

        assertEquals(3, tracker.updates.size());
        assertEquals(18, tracker.updates.get(0).temperature, 0.01);
        assertEquals(22, tracker.updates.get(1).temperature, 0.01);
        assertEquals(25, tracker.updates.get(2).temperature, 0.01);
    }

    @Test
    public void testUnsubscribeMidStreamStopsUpdates() {
        station.registerObserver(tracker);

        station.setWeatherData(18, 45, 1015, 10);
        assertEquals(1, tracker.updates.size());

        station.removeObserver(tracker);

        station.setWeatherData(22, 60, 1010, 8);
        assertEquals(1, tracker.updates.size()); // No new update

        station.registerObserver(tracker);

        station.setWeatherData(25, 70, 1015, 12);
        assertEquals(2, tracker.updates.size()); // Resumes after re-subscribe
    }

    @Test
    public void testAllObserversSeesSameCurrentReading() {
        TrackingObserver tracker1 = new TrackingObserver();
        TrackingObserver tracker2 = new TrackingObserver();
        TrackingObserver tracker3 = new TrackingObserver();

        station.registerObserver(tracker1);
        station.registerObserver(tracker2);
        station.registerObserver(tracker3);

        station.setWeatherData(25, 70, 1015, 15);

        TrackingObserver.Update u1 = tracker1.updates.get(0);
        TrackingObserver.Update u2 = tracker2.updates.get(0);
        TrackingObserver.Update u3 = tracker3.updates.get(0);

        assertEquals(u1.temperature, u2.temperature, 0.01);
        assertEquals(u2.temperature, u3.temperature, 0.01);
        assertEquals(u1.humidity, u2.humidity, 0.01);
        assertEquals(u1.pressure, u2.pressure, 0.01);
    }

    @Test
    public void testReadingHistoryRemainConsistentAcrossObserverLifespan() {
        TrackingObserver tracker1 = new TrackingObserver();

        station.registerObserver(tracker1);
        station.setWeatherData(18, 45, 1015, 10);
        station.setWeatherData(22, 60, 1010, 8);

        // Get history at first observation point
        List<WeatherData> history1 = station.getReadings();

        station.setWeatherData(25, 70, 1015, 12);

        // Get history after another update
        List<WeatherData> history2 = station.getReadings();

        assertTrue(history1.size() < history2.size());
        assertEquals(history1.get(0).temperature, history2.get(0).temperature, 0.01);
        assertEquals(history1.get(1).temperature, history2.get(1).temperature, 0.01);
    }

    @Test
    public void testWorksWithMixedObserverSubscribeUnsubscribeSequence() {
        TrackingObserver tracker1 = new TrackingObserver();
        TrackingObserver tracker2 = new TrackingObserver();
        TrackingObserver tracker3 = new TrackingObserver();

        // Day 1: Only tracker1
        station.registerObserver(tracker1);
        station.setWeatherData(18, 45, 1015, 10);
        assertEquals(1, tracker1.updates.size());
        assertEquals(0, tracker2.updates.size());
        assertEquals(0, tracker3.updates.size());

        // Day 2: Add tracker2
        station.registerObserver(tracker2);
        station.setWeatherData(22, 60, 1010, 8);
        assertEquals(2, tracker1.updates.size());
        assertEquals(1, tracker2.updates.size());
        assertEquals(0, tracker3.updates.size());

        // Day 3: Add tracker3, remove tracker1
        station.registerObserver(tracker3);
        station.removeObserver(tracker1);
        station.setWeatherData(25, 70, 1015, 12);
        assertEquals(2, tracker1.updates.size()); // Stopped
        assertEquals(2, tracker2.updates.size());
        assertEquals(1, tracker3.updates.size());

        // Day 4: Re-add tracker1
        station.registerObserver(tracker1);
        station.setWeatherData(28, 75, 1020, 14);
        assertEquals(3, tracker1.updates.size()); // Resumed
        assertEquals(3, tracker2.updates.size());
        assertEquals(2, tracker3.updates.size());
    }

    @Test
    public void testRespectsLast10ReadingsHistoryLimitAcrossObserverLifecycle() {
        TrackingObserver tracker = new TrackingObserver();
        station.registerObserver(tracker);

        // Generate 15 readings
        for (int i = 0; i < 15; i++) {
            station.setWeatherData(10 + i, 50 + i, 1010 + i, 10);
        }

        // Should have exactly 10 readings in history
        List<WeatherData> readings = station.getReadings();
        assertEquals(10, readings.size());

        // Should be the last 10 updates
        assertEquals(15, readings.get(0).temperature, 0.01); // 10 + 5
        assertEquals(24, readings.get(9).temperature, 0.01); // 10 + 14
    }

    @Test
    public void testSupportsArbitraryObserverImplementations() {
        class TempAverageObserver implements Observer {
            double tempSum = 0;
            int count = 0;

            @Override
            public void update(WeatherStation station) {
                WeatherData current = station.getCurrentReading();
                tempSum += current.temperature;
                count++;
            }

            double getAverage() {
                return count > 0 ? tempSum / count : 0;
            }
        }

        TempAverageObserver avgObserver = new TempAverageObserver();
        station.registerObserver(avgObserver);

        station.setWeatherData(20, 60, 1010, 10);
        station.setWeatherData(24, 65, 1015, 12);
        station.setWeatherData(26, 70, 1020, 15);

        double avg = avgObserver.getAverage();
        assertEquals((20 + 24 + 26) / 3.0, avg, 0.01);
    }

    @Test
    public void testAllThreeConcreteObserversCoexistWithoutInterference() {
        TrackingObserver tracker = new TrackingObserver();

        station.registerObserver(tempChart);
        station.registerObserver(humidityChart);
        station.registerObserver(pressureChart);
        station.registerObserver(tracker);

        // Perform updates
        station.setWeatherData(18, 45, 1015, 10);
        station.setWeatherData(25, 70, 1020, 15);

        // All should have been notified
        assertEquals(2, tracker.updates.size());

        // Reading history accessible to all
        List<WeatherData> readings = station.getReadings();
        assertTrue(readings.size() > 2);
    }
}

// Tracking Observer (package-private, not a test class)
class TrackingObserver implements Observer {
    static class Update {
        double temperature, humidity, pressure, windSpeed;
        Update(double t, double h, double p, double w) {
            temperature = t;
            humidity = h;
            pressure = p;
            windSpeed = w;
        }
    }

    List<Update> updates = new ArrayList<>();

    @Override
    public void update(WeatherStation station) {
        WeatherData reading = station.getCurrentReading();
        updates.add(new Update(reading.temperature, reading.humidity, reading.pressure, reading.windSpeed));
    }
}
